frontend/lib/src/components/shared/StreamlitMarkdown/Heading.tsx

```diff
@@ -54,63 +55,78 @@ function makeMarkdownHeading(tag: string, markdown: string): string {
-                  }}
-                />
-                <InlineTooltipIcon content={help} />
-              </StyledLabelHelpWrapper>
-            ) : (
-              <RenderedMarkdown
-                source={makeMarkdownHeading(tag, heading)}
-                allowHTML={false}
-                // this is purely an inline string
-                overrideComponents={{
-                  p: Fragment,
-                  h1: Fragment,
-                  h2: Fragment,
-                  h3: Fragment,
-                  h4: Fragment,
-                  h5: Fragment,
-                  h6: Fragment,
-                }}
-              />
-            )}
-          </HeadingWithAnchor>
-        </StyledHeaderContainer>
-        {/* Only the first line of the body is used as a heading, the remaining text is added as regular mardkown below. */}
-        {rest.length > 0 && (
-          <RenderedMarkdown source={rest.join("\n")} allowHTML={false} />
-        )}
-      </StyledStreamlitMarkdown>
+    <div className="stHeadingContainer">
+      <div className="stMarkdown" style={{ width }}>
+        <StyledStreamlitMarkdown
+          isCaption={Boolean(false)}
+          isInSidebar={isSidebar}
+          style={{ width }}
+          data-testid="stMarkdownContainer"
+        >
+          <StyledHeaderContainer>
+            <HeadingWithAnchor
+              tag={tag}
+              anchor={anchor}
+              hideAnchor={hideAnchor}
+            >
+              {help ? (
+                <StyledLabelHelpWrapper>
+                  <RenderedMarkdown
+                    source={makeMarkdownHeading(tag, heading)}
+                    allowHTML={false}
+                    // this is purely an inline string
+                    overrideComponents={{
+                      p: Fragment,
+                      h1: Fragment,
+                      h2: Fragment,
+                      h3: Fragment,
+                      h4: Fragment,
+                      h5: Fragment,
+                      h6: Fragment,
+                    }}
+                  />
+                  <InlineTooltipIcon content={help} />
+                </StyledLabelHelpWrapper>
+              ) : (
+                <>
+                  <RenderedMarkdown
+                    source={makeMarkdownHeading(tag, heading)}
+                    allowHTML={false}
+                    // this is purely an inline string
+                    overrideComponents={{
+                      p: Fragment,
+                      h1: Fragment,
+                      h2: Fragment,
+                      h3: Fragment,
+                      h4: Fragment,
+                      h5: Fragment,
+                      h6: Fragment,
+                    }}
+                  />
+                </>
+              )}
+            </HeadingWithAnchor>
+          </StyledHeaderContainer>
+          {/* Only the first line of the body is used as a heading, the remaining text is added as regular mardkown below. */}
```

Comment:
Maybe a bit unrelated, but the spec mentions that it might be nice to have a way to add a caption under the header. Maybe we can actually render this remaining text via the caption styling? Probably something to double check with product.